switchtec.c

```diff
@@ -251,12 +262,24 @@ static void mrpc_event_work(struct work_struct *work)
 static void mrpc_timeout_work(struct work_struct *work)
 {
 	struct switchtec_dev *stdev;
+	struct switchtec_user *stuser;
 	u32 status;
 
 	stdev = container_of(work, struct switchtec_dev, mrpc_timeout.work);
 
 	dev_dbg(&stdev->dev, "%s\n", __func__);
 
+	if (!check_access(stdev)) {
+		if (list_empty(&stdev->mrpc_queue))
+			return;
+
+		stuser = list_entry(stdev->mrpc_queue.next,
+				    struct switchtec_user, list);
+
+		stuser_set_state(stuser, MRPC_ERROR);
+		goto complete_and_out;
```

Comment:
Also you're now calling mrpc_complete_cmd which requires the mutex to be held...